docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,302 @@
+
+.. code-block:: php
+
+   <?php
+
+   $collection = (new MongoDB\Client())->selectCollection('test', 'person', ['codec' => new PersonCodec()]);
+
+   $person = new Person('Jane Doe');
+   $collection->insertOne($person);
+
+   $person = $collection->findOne();
+
+The example above selects a collection and instructs it to use the ``PersonCodec`` for encoding and decoding documents.
+When inserting data, the ``PersonCodec`` is used to encode the document. When retrieving data, the ``PersonCodec`` is
+used to decode BSON data into a ``Person`` instance.
+
+.. note::
+
+   When working on a collection with a codec, all incoming and outgoing data will be handled with the given codec. This
+   affects the following methods: ``aggregate``, ``bulkWrite``, ``find``, ``findOne``, ``findOneAndDelete``,
+   ``findOneAndReplace``, ``findOneAndUpdate``, ``insertMany``, ``insertOne``, and ``replaceOne``. To disable the codec
+   for a specific operation, you can provide ``null`` for the ``codec`` option in the operation.
+
+Generic Codecs
+--------------
+
+The previous example showed how to define a codec for a specific class. However, sometimes you want to define codecs to
+handle a given type in all documents. For such codecs, you can implement the ``MongoDB\Codec\Codec`` interface. The
+following example defines a codec to store ``DateTimeInterface`` instances as BSON dates including the timezone:
+
+.. code-block:: php
+
+   <?php
+
+   final class DateTimeCodec implements MongoDB\Codec\Codec
+   {
+       // These traits define commonly used functionality to avoid duplication
+       use MongoDB\Codec\DecodeIfSupported;
+       use MongoDB\Codec\EncodeIfSupported;
+
+       public function canDecode($value): bool
+       {
+           // For maximum compatibility, this codec supports decoding both UTCDateTime instances and documents with a
+           // UTCDateTime instance and optional timezone
+           return
+               $value instanceof MongoDB\BSON\UTCDateTime ||
+               $value instanceof MongoDB\BSON\Document && $value->has('utc');
+       }
+
+       public function canEncode($value): bool
+       {
+           return $value instanceof DateTimeInterface;
+       }
+
+       public function decode($value): DateTimeImmutable
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $utc = $value instanceof MongoDB\BSON\UTCDateTime
+               ? $value
+               : $value->get('utc');
+
+           if (! $utc instanceof UTCDateTime) {
+               throw UnsupportedValueException::invalidDecodableValue($utc);
+           }
+
+           $dateTime = $utc->toDateTime();
+
+           if ($value instanceof MongoDB\BSON\Document && $value->has('tz')) {
+               $dateTime->setTimeZone(new DateTimeZone($value->get('tz')));
+           }
+
+           return DateTimeImmutable::fromMutable($dateTime);
```

Comment:
That make me think that `UTCDateTime::toDateTime` should return a `DateTimeImmutable`. Something for 2.0?

Reply:
I'm not sure if we've previously discussed this, but I do think it's worth a ticket for 2.0 - either for changing the return type of `toDateTime` or to add a separate method.

Reply:
Created [PHPC-2286](https://jira.mongodb.org/browse/PHPC-2286)